// @sinclair/typebox/build/cjs/type/record/record.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.Record = Record;
const type_1 = require("../create/type");
const index_1 = require("../computed/index");
const index_2 = require("../object/index");
const index_3 = require("../never/index");
const index_4 = require("../union/index");
const index_5 = require("../template-literal/index");
const index_6 = require("../patterns/index");
const index_7 = require("../indexed/index");
const index_8 = require("../symbols/index");
// ------------------------------------------------------------------
// ValueGuard
// ------------------------------------------------------------------
const value_1 = require("../guard/value");
// ------------------------------------------------------------------
// TypeGuard
// ------------------------------------------------------------------
const kind_1 = require("../guard/kind");
// ------------------------------------------------------------------
// RecordCreateFromPattern
// ------------------------------------------------------------------
// prettier-ignore
function RecordCreateFromPattern(pattern, T, options) {
    return (0, type_1.CreateType)({
        [index_8.Kind]: 'Record',
        type: 'object',
        patternProperties: { [pattern]: T }
    }, options);
}
// ------------------------------------------------------------------
// RecordCreateFromKeys
// ------------------------------------------------------------------
// prettier-ignore
function RecordCreateFromKeys(K, T, options) {
    const Acc = {};
    for (const K2 of K)
        Acc[K2] = T;
    return (0, index_2.Object)(Acc, { ...options, [index_8.Hint]: 'Record' });
}
// prettier-ignore
function FromTemplateLiteralKey(K, T, options) {
    return ((0, index_5.IsTemplateLiteralFinite)(K)
        ? RecordCreateFromKeys((0, index_7.IndexPropertyKeys)(K), T, options)
        : RecordCreateFromPattern(K.pattern, T, options));
}
// prettier-ignore
function FromUnionKey(K, T, options) {
    return RecordCreateFromKeys((0, index_7.IndexPropertyKeys)((0, index_4.Union)(K)), T, options);
}
// prettier-ignore
function FromLiteralKey(K, T, options) {
    return RecordCreateFromKeys([K.toString()], T, options);
}
// prettier-ignore
function FromRegExpKey(K, T, options) {
    return RecordCreateFromPattern(K.source, T, options);
}
// prettier-ignore
function FromStringKey(K, T, options) {
    const pattern = (0, value_1.IsUndefined)(K.pattern) ? index_6.PatternStringExact : K.pattern;
    return RecordCreateFromPattern(pattern, T, options);
}
// prettier-ignore
function FromAnyKey(K, T, options) {
    return RecordCreateFromPattern(index_6.PatternStringExact, T, options);
}
// prettier-ignore
function FromNeverKey(K, T, options) {
    return RecordCreateFromPattern(index_6.PatternNeverExact, T, options);
}
// prettier-ignore
function FromIntegerKey(_, T, options) {
    return RecordCreateFromPattern(index_6.PatternNumberExact, T, options);
}
// prettier-ignore
function FromNumberKey(_, T, options) {
    return RecordCreateFromPattern(index_6.PatternNumberExact, T, options);
}
// ------------------------------------------------------------------
// TRecordOrObject
// ------------------------------------------------------------------
/** `[Json]` Creates a Record type */
function Record(key, type, options = {}) {
    // prettier-ignore
    return ((0, kind_1.IsRef)(type) ? (0, index_1.Computed)('Record', [key, type]) :
        (0, kind_1.IsRef)(key) ? (0, index_1.Computed)('Record', [key, type]) :
            (0, kind_1.IsUnion)(key) ? FromUnionKey(key.anyOf, type, options) :
                (0, kind_1.IsTemplateLiteral)(key) ? FromTemplateLiteralKey(key, type, options) :
                    (0, kind_1.IsLiteral)(key) ? FromLiteralKey(key.const, type, options) :
                        (0, kind_1.IsInteger)(key) ? FromIntegerKey(key, type, options) :
                            (0, kind_1.IsNumber)(key) ? FromNumberKey(key, type, options) :
                                (0, kind_1.IsRegExp)(key) ? FromRegExpKey(key, type, options) :
                                    (0, kind_1.IsString)(key) ? FromStringKey(key, type, options) :
                                        (0, kind_1.IsAny)(key) ? FromAnyKey(key, type, options) :
                                            (0, kind_1.IsNever)(key) ? FromNeverKey(key, type, options) :
                                                (0, index_3.Never)(options));
}
